src/spec-node/policy.ts

```diff
@@ -0,0 +1,178 @@
+ *--------------------------------------------------------------------------------------------*/
+
+import * as jsonc from 'jsonc-parser';
+
+import { LifecycleCommand } from '../spec-common/injectHeadless';
+import { DevContainerFromDockerComposeConfig, DevContainerFromDockerfileConfig, DevContainerFromImageConfig } from '../spec-configuration/configuration';
+import { ImageMetadataEntry } from './imageMetadata';
+import { CLIHost } from '../spec-common/cliHost';
+import { Log, LogLevel } from '../spec-utils/log';
+import { ContainerError } from '../spec-common/errors';
+
+export type PolicyConstraints = Constraint[];
+
+const readPolicyFileError = 'Failed to parse policy constraints file';
+const errorPrefix = 'Policy violation';
+
+export interface Constraint {
+	action: 'filter' | 'deny';    	// | 'transformation';
+	selector: string; 				// TODO: Define interface for selector
+	// value?: any;  				// TODO: Allow requiring a match value as well for more complex constraints
+}
+
+export async function readPolicyConstraintsFromFile(params: { output?: Log; policyFile?: string; cliHost: CLIHost }): Promise<PolicyConstraints | undefined> {
+	const { policyFile, cliHost, output } = params;
+	if (!policyFile) {
+		return;
+	}
+
+	try {
+		const fileBuff = await cliHost.readFile(policyFile);
+		const parseErrors: jsonc.ParseError[] = [];
+		const policyConstraints = jsonc.parse(fileBuff.toString(), parseErrors) as PolicyConstraints;
+		if (parseErrors.length) {
+			// Log errors
+			if (output) {
+				parseErrors.forEach(e => {
+					output.write(`(${e.offset}) ${e.error}`, LogLevel.Error);
+				});
+			}
+			throw new Error(`Invalid json data`);
+		}
+
+		// Validate that the top-level item is an array
+		if (!Array.isArray(policyConstraints)) {
+			throw new Error(`Policy constraints file must be an array of constraint objects`);
+		}
+
+		return policyConstraints;
+	}
+	catch (e) {
+		throw new ContainerError({
+			description: readPolicyFileError,
+			originalError: e
+		});
+	}
+}
+
+export function applyConstraintsToMetadataEntries(params: { output: Log }, metadata: ImageMetadataEntry[], constraints: PolicyConstraints | undefined) {
+	if (!constraints) {
+		return metadata;
+	}
+	const copy = metadata.slice();
+	return copy.map(entry => apply<ImageMetadataEntry>(params, entry, constraints));
+}
+
+export function applyConstraintsToSingleContainerConfig(params: { output: Log }, config: DevContainerFromImageConfig | DevContainerFromDockerfileConfig, constraints: PolicyConstraints | undefined) {
+	if (!constraints) {
+		return config;
+	}
+	const configApplied = apply(params, config, constraints);
+	return {
+		...configApplied,
+		runArgs: applyConstraintsToRunArgs(params, configApplied.runArgs, constraints)
+	};
+}
+
+export function applyConstraintsToComposeConfig(params: { output: Log }, config: DevContainerFromDockerComposeConfig, constraints: PolicyConstraints | undefined) {
```

Comment:
For a `docker-compose` dev container, one can still add properties like `privileged` or `capAdd` to the `.devcontainer.json`. How is that being handled? 🤔